mlir/lib/Conversion/UBToEmitC/UBToEmitC.cpp

```diff
@@ -0,0 +1,85 @@
+//===- UBToEmitC.cpp - UB to EmitC dialect conversion ---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/UBToEmitC/UBToEmitC.h"
+
+#include "mlir/Dialect/EmitC/IR/EmitC.h"
+#include "mlir/Dialect/EmitC/Transforms/TypeConversions.h"
+#include "mlir/Dialect/UB/IR/UBOps.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_CONVERTUBTOEMITC
+#include "mlir/Conversion/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+
+namespace {
+struct PoisonOpLowering : public OpConversionPattern<ub::PoisonOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ub::PoisonOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    const TypeConverter *converter = getTypeConverter();
+    Type convertedType = converter->convertType(op.getType());
+
+    if (!convertedType)
+      return rewriter.notifyMatchFailure(op.getLoc(), "type conversion failed");
+
+    Attribute value;
+    if (emitc::isIntegerIndexOrOpaqueType(convertedType)) {
+      value = IntegerAttr::get((emitc::isPointerWideType(convertedType)
+                                    ? IndexType::get(getContext())
+                                    : convertedType),
+                               0);
+    } else if (emitc::isSupportedFloatType(convertedType)) {
+      value = FloatAttr::get(convertedType, 0);
+    } else {
+      return rewriter.notifyMatchFailure(
+          op.getLoc(), "only scalar poison values can be lowered");
+    }
+
+    // Any constant will be fine to lower a poison op
+    rewriter.replaceOpWithNewOp<emitc::VariableOp>(op, convertedType, value);
```

Comment:
We may not even need variables, constants should be fine, let me have a fresh look at this next week.